packages/@aws-cdk/cx-api/lib/private/flag-modeling.ts

```diff
@@ -0,0 +1,39 @@
+export enum FlagType {
+  /**
+   * Change the default behavior of the API
+   *
+   * The old behavior is not disrecommended, and possible to achieve with source
+   * code changes. Also valid for changes that don't affect CloudFormation, but
+   * the CXAPI contract.
+   */
+  ApiDefault,
+
+  /**
+   * Address a bug/introduce a recommended change
+   *
+   * The old behavior is no longer recommended. The only way to achieve it is by
+   * keeping the flag at the legacy value.
+   */
+  BugFix,
+
+  /**
+   * Advertise the presence of this context option in `cdk.json`
+   */
+  VisibleContext,
+};
+
+export interface FlagInfo {
+  /** Flag type */
+  readonly type: FlagType;
+  /** Single-line description for the flag */
+  readonly summary: string;
+  /** Detailed description for the flag */
+  readonly details: string;
+  /** Version number the flag was introduced in each version line */
+  readonly introducedIn: { v1?: string; v2?: string };
+  /** Default value in the given major version */
```

Comment:
I'm having trouble thinking of a better name (`permanent`, `immutable`, `stable` ?), but first thought reading this was that this is the default, but overridable value in `v2`. But these are things that you can't change in `v2` right? Maybe just more in the docstring would be fine?

Reply:
It's not exactly.

* `default` indicates what the default value is, if no override is supplied. It's *also* true that these can't be changed anymore, but that's a fact caused by the lifecycle of v2, moreso than implied by the semantics of the `default` value.
* `introducedIn` has the semantics you mean: if there's no `introducedIn.v2` entry, then the flag is not configurable in v2.

Reply:
> default indicates what the default value is, if no override is supplied. It's also true that these can't be changed anymore, but that's a fact caused by the lifecycle of v2, moreso than implied by the semantics of the default value.

Could they ever be changed though? Going forward we would never add a new flag _and_ set `defaults` until we have a `v3` at which point we would do `defaults: { v3: true } `. But even then for `v3` you wouldn't be able to override it right? 

> introducedIn has the semantics you mean (if that's what you mean): if there's no introducedIn.v2 entry, then the flag is not configurable in v2.

I actually wasn't clear on this, which makes me think it should be named something like `configurableIn`.

I'm not actually opposed to the current naming, your explanations clear things up for me which makes me think they need to go in the docstrings 😃

Reply:
> Could they ever be changed though? Going forward we would never add a new flag and set defaults until we have a v3 at which point we would do defaults: { v3: true } . But even then for v3 you wouldn't be able to override it right?

Presumably we would have an alpha period for v3, during which we would be allowed to change the defaults.

Also, just realized we SHOULD technically be allowed to change these, in order to flip the meaning of negative flags:

* `disableAutomaticRoles`: `{ default: false, recommendedValue: true }`

➡️➡️➡️

* `automaticRoles`: `{ default: true, recommendedValue: false }`

Double negatives are generally hard to read so adding a `default: true` would allow us to bypass that.

I currently have a test that asserts that the set of `defaults` doesn't change, though.

Reply:
Added some documentation